### ISSUE TITLE
Limit option when creating new data sources breaks SQL logic

### ISSUE THREAD

**Issue by romanmotsar** [cognitedata/cdp-spark-datasource#166]
When we are creating a table with `cdp_spark_datasource` and specifying limit option we expect, that all queries will be limited to option if there are more data 

```
%sql

-- DROP TABLE IF EXISTS kjoereplan.api_events
-- CREATE TABLE IF NOT EXISTS kjoereplan.api_events using com.cognite.spark.datasource
-- options (type 'events',
--          apiKey '%s',
--          project 'akerbp',
--          limit '1000000',
--          batchSize '50000')

select count(*) from kjoereplan.api_events;
```
Output: 1000000

But, the actual result is that we are applying a filter on limited data (in the example is 1M records)
```
%sql
select count(*) from kjoereplan.api_events where type=='Workorder'
```
Output: 4

Expected result should be the same as described below, but not used for now because of performance issues
```
%sql

-- DROP TABLE IF EXISTS kjoereplan.api_events_test_no_limit
-- CREATE TABLE IF NOT EXISTS kjoereplan.api_events_test_no_limit using com.cognite.spark.datasource
-- options (type 'events',
--          apiKey '%s',
--          project 'akerbp',
--          batchSize '50000')

select count(*) from kjoereplan.api_events_test_no_limit where type=='Workorder' LIMIT 1000000;
```
Output: 1000000

By analyzing SQL query we can find that firstly we have prefetched 1M records, and then we are applying a filter, and expected result will be fetched from filtered data 
![image](https://user-images.githubusercontent.com/47787026/53870920-709a1c80-3ffb-11e9-9f60-9661cb9af88e.png)

So, it seems like filtering and fetching stages are swapped, and it's break an common SQL logic

**Comment by wjoel** [comment]
Thank you for reporting this.

This is a documentation bug: `limit` specifies the maximum number of items to be fetched from the Cognite Data Platform, not the number of items to be returned in the end (ie. after filtering). It's not the same as SQLs `LIMIT 100...0` when there's filtering etc.

The reason it exists is because the Spark Data Source API doesn't do push down of the values in the `LIMIT 100...0` clause, so we can't see that value, and thus must fetch all items. To work around this we instead allow the user to specify `.option("limit", "100...0")` to limit the number of records fetched from CDP, but in retrospect calling it `"limit"` is confusing, because it's (subtly) different.

In other words, your results are expected but surprising. It should be documented and better explained. The data source does have to fetch all items from CDP to do your filtering correctly, since we don't (currently) push down the filters on `type`, and can unfortunately not use `.option("limit", ...)` for this use case.

**Comment by jaketbouma** [comment]
Interesting! Thanks for the explanation @wjoel , makes sense.

I'll be the first to say that the pushdown of `granularity` etc for the datapoints resource is magic.
Is a similar kind of magic possible to push down the parameters for https://doc.cognitedata.com/api/0.5/#operation/getEvents so that the `WHERE TYPE='abc'` is pushed into CDP.

This is needed to get acceptable performance in a couple of use cases. For now, we will have to cache all events.

**Comment by romanmotsar** [comment]
Oh, thanks for explanation @wjoel !

Join to @jaketbouma, pushing arguments down (especially filtering) to Cognite Data Platform before fetching will be very useful and may resolve many performance issues with retrieving and filtering very specific data from CDP

**Comment by wjoel** [comment]
I've created https://github.com/cognitedata/cdp-spark-datasource/issues/172 for the filter push down for events. The documentation was improved in https://github.com/cognitedata/cdp-spark-datasource/pull/169.
